### ISSUE TITLE
Setting analyzer on custom index doesn't work

### ISSUE THREAD

**Issue by bjarnef** [umbraco/Umbraco-CMS#14054]
### Which Umbraco version are you using? (Please write the *exact* version, example: 10.1.0)

10.4.0

### Bug summary

In a working project on Umbraco v10.4.0 we have a custom Examine index, where we have configuration of fields to be sortable.
Furthermore we tried to replace/overwrite the default analyzer here, but it doesn't seem to work as Examine dashboard still shows StandardAnalyzer.

We tried a few different analyzers like `CultureInvariantWhitespaceAnalyzer` and `KeywordAnalyzer`.

```
public sealed class ConfigureCourseInstanceIndexOptions : IConfigureNamedOptions<LuceneDirectoryIndexOptions>
{
	private readonly IUmbracoIndexConfig _umbracoIndexConfig;
	private readonly IOptions<IndexCreatorSettings> _settings;

	public ConfigureCourseInstanceIndexOptions(
		IUmbracoIndexConfig umbracoIndexConfig,
		IOptions<IndexCreatorSettings> settings)
	{
		_umbracoIndexConfig = umbracoIndexConfig;
		_settings = settings;
	}

	public void Configure(string name, LuceneDirectoryIndexOptions options)
	{
		switch (name)
		{
			case Constants.Examine.Indexes.CourseInstanceIndexName:
				options.Analyzer = new CultureInvariantWhitespaceAnalyzer();
				options.FieldDefinitions.TryAdd(new FieldDefinition(UmbracoExamineFieldNames.NodeNameFieldName, FieldDefinitionTypes.FullTextSortable));
				options.FieldDefinitions.TryAdd(new FieldDefinition(Constants.Examine.Course.FieldNames.CourseName, FieldDefinitionTypes.FullTextSortable));
				options.FieldDefinitions.TryAdd(new FieldDefinition(Constants.Examine.CourseInstance.FieldNames.StartDate, FieldDefinitionTypes.DateTime));
				options.FieldDefinitions.TryAdd(new FieldDefinition(Constants.Examine.CourseInstance.FieldNames.LatestValidDate, FieldDefinitionTypes.DateTime));
				options.FieldDefinitions.TryAdd(new FieldDefinition(Constants.Examine.CourseInstance.FieldNames.RegistrationDeadline, FieldDefinitionTypes.DateTime));
				options.FieldDefinitions.TryAdd(new FieldDefinition(Constants.Examine.CourseInstance.FieldNames.EndDate, FieldDefinitionTypes.DateTime));
				options.FieldDefinitions.TryAdd(new FieldDefinition(Constants.Examine.CourseInstance.FieldNames.AvailableSeats, FieldDefinitionTypes.Integer));
				options.FieldDefinitions.TryAdd(new FieldDefinition(Constants.Examine.CourseInstance.FieldNames.HasSeats, FieldDefinitionTypes.Integer));
				options.FieldDefinitions.TryAdd(new FieldDefinition(Constants.Examine.CourseInstance.FieldNames.Duration, FieldDefinitionTypes.Double));
				options.FieldDefinitions.TryAdd(new FieldDefinition(Constants.Examine.Course.FieldNames.Duration, FieldDefinitionTypes.Double));
				options.FieldDefinitions.TryAdd(new FieldDefinition(Constants.Examine.Course.FieldNames.HasInstances, FieldDefinitionTypes.Integer));  
				break;
		}
	}

	public void Configure(LuceneDirectoryIndexOptions options)
		=> Configure(string.Empty, options);
}
```

![image](https://user-images.githubusercontent.com/2919859/229711281-f4df1330-5cba-42b4-9edc-386db76f9708.png)

This is the same approach used in core here:
https://github.com/umbraco/Umbraco-CMS/blob/33adbf41fa1f5c5d0759c70a7116114107addf56/src/Umbraco.Examine.Lucene/DependencyInjection/ConfigureIndexOptions.cs#L32-L46

A colleague of mine tried to set this outside the switch case and noticed it did change the Analyzer on e.g. `ExternalIndex`.

### Steps to reproduce

Add a custom Examine index + configuration of this.
In the configuration set a different analyzer than StandardAnalyzer.

---
_This item has been added to our backlog [AB#28447](https://dev.azure.com/umbraco/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/28447)_

**Comment by bjarnef** [comment]
Okay, I figured out this works after changing from this:


```
public class CourseInstanceIndexComposer : IComposer
{
	public void Compose(IUmbracoBuilder builder)
	{
		// Custom Examine configuration
		builder.Services.ConfigureOptions<ConfigureCourseInstanceIndexOptions>();

		builder.Services.AddExamineLuceneIndex<CourseInstanceIndex, ConfigurationEnabledDirectoryFactory>(Constants.Examine.Indexes.CourseInstanceIndexName);

		builder.Services.AddSingleton<IValueSetBuilder<CourseDataIndexingModel>>(factory =>
			new CourseInstanceValueSetBuilder(factory.GetRequiredService<IUmbracoContextFactory>()));

		builder.Services.AddSingleton<IIndexPopulator>(factory => new CourseInstanceIndexPopulator(
			factory.GetRequiredService<ICourseDataService>(),
			factory.GetRequiredService<IScopeProvider>(),
			factory.GetRequiredService<IValueSetBuilder<CourseDataIndexingModel>>()));
	}
}
```

```
public class CourseInstanceIndexComposer : IComposer
{
	public void Compose(IUmbracoBuilder builder)
	{
		builder.Services.AddSingleton<IValueSetBuilder<CourseDataIndexingModel>>(factory =>
			new CourseInstanceValueSetBuilder(factory.GetRequiredService<IUmbracoContextFactory>()));

		builder.Services.AddExamineLuceneIndex<CourseInstanceIndex, ConfigurationEnabledDirectoryFactory>(Constants.Examine.Indexes.CourseInstanceIndexName)
			.ConfigureOptions<ConfigureCourseInstanceIndexOptions>();

		builder.Services.AddSingleton<IIndexPopulator>(factory => new CourseInstanceIndexPopulator(
			factory.GetRequiredService<ICourseDataService>(),
			factory.GetRequiredService<IScopeProvider>(),
			factory.GetRequiredService<IValueSetBuilder<CourseDataIndexingModel>>()));
	}
}
```

![image](https://user-images.githubusercontent.com/2919859/229757334-a0aa3149-923f-4934-892d-07b2de890510.png)

This is also how it is handled in core here:
https://github.com/umbraco/Umbraco-CMS/blob/33adbf41fa1f5c5d0759c70a7116114107addf56/src/Umbraco.Examine.Lucene/DependencyInjection/UmbracoBuilderExtensions.cs#L33

**Comment by bjarnef** [comment]
@bergmania @Zeegaan should it be possible to register the configuration before?

**Comment by kjac** [comment]
Hi @bjarnef,

There are quite a few factors that work against us being able to do the registrations in an arbitrary order. For now we need to settle for documenting this.

On that note we're looking at the custom Examine index docs again, as they are somewhat misleading or insufficient at the moment.

Out of curiosity (and because it impacts said docs rewrite), how do you handle index re-population when content changes in your course index?

**Comment by bjarnef** [comment]
Hi @kjac 

Okay, for now we need to ensure `.ConfigureOptions()` is called after the Examine index has been registred.

Regarding re-indexing I tried two approaches (both work).
https://github.com/umbraco/UmbracoDocs/issues/5003#issuecomment-1486546652

1. Implementing `IUmbracoIndexingHandler` as in core, but it required copying much of the existing stuff from Umbraco core to start a background task. It also required implementing methods for media and member, which I didn't need in this case, but it was running in background and didn't block publishing notification.

2. Use the approach from [Examine PDF](https://github.com/umbraco/UmbracoExamine.PDF) repository adding `AddToIndex` and `RemoveFromIndex` methods in the populater. This works as work, but if fetching (part) of the index data comes from internal/external data source, it seems it has to wait until publish notification is shown. We didn't optimize this yet to fetch specific items yet, so I just tested `GetAll()` and LINQ query on the result to filter a specific result (same as in 1. using background task).
